src/interface/lazy.jl

```diff
@@ -147,9 +150,50 @@ function tensordot(A::LazyTensor{T1, N1}, B::LazyTensor{T2, N2}, idxs; mult_op=*
     AB = mapjoin(immediate(mult_op), relabel(A.data, A_fields), relabel(B.data, B_fields))
     AB_reduce = aggregate(immediate(add_op), immediate(init), AB, reduce_fields...)
     S = fixpoint_type(add_op, init, AB_reduce)
-    return LazyTensor{S}(identify(AB_reduce), extrude)
+    return LazyTensor{S}(identify(AB_reduce), extrude, init)
 end
 
+#=
+# einsum represents an expression of the form Out[`out_idxs`] += ∏ F[F_idxs]
+# where each entry in args is a (lazy) tensor and a set of indices (F, F_idxs)
+function einsum(out_idxs, args...; add_op=+, mult_op=*, init = initial_value(add_op, Float64))
+    if any([length(arg) != 2 for arg in args])
+        throw(ArgumentError("einsum arguments must be a tuple of length 2 `(T::LazyTensor, Idxs::Tuple{String,...})`"))
+    end
+
+    if any([ length(arg[2]) != length(arg[1].extrude) for arg in args])
+        throw(ArgumentError("einsum arguments must have an equal number of dimensions and named indices"))
+    end
+    args = [(arg[1], [string(idx) for idx in arg[2]]) for arg in args]
+    # For some reason, union automatically translates single character strings to type `Char`
+    s_idxs::Vector{String} = [string(idx) for idx in union(flatten([arg[2] for arg in args])...)]
+    s_to_idx = Dict(s_idx => field(gensym(Symbol(s_idx))) for s_idx in s_idxs)
+    factors = []
+    for arg in args
+        push!(factors, relabel(arg[1].data, [s_to_idx[s_idx] for s_idx in arg[2]]...))
+    end
+    out_fields = [s_to_idx[s_idx] for s_idx in out_idxs]
+    reduce_fields = [s_to_idx[s_idx] for s_idx in s_idxs if !(s_idx in out_idxs)]
+
+    extrude = []
+    for idx in out_idxs
+        is_extruded = true
+        for arg in args
+            idx_pos = findfirst(isequal(idx), arg[2])
+            isnothing(idx_pos) && continue
+            is_extruded = is_extruded && arg[1].extrude[idx_pos]
+        end
+        push!(extrude, is_extruded)
+    end
+    out_fields_in_input_order = [s_to_idx[s_idx] for s_idx in s_idxs if s_idx in out_idxs]
+    t_prod = mapjoin(immediate(mult_op), factors...)
+    t_sum = aggregate(immediate(add_op), immediate(init), t_prod, reduce_fields...)
+    t_ordered = reorder(relabel(t_sum, out_fields_in_input_order...), out_fields...)
+    S = fixpoint_type(add_op, init, t_ordered)
+    return LazyTensor{S}(identify(t_ordered), Tuple(extrude), init)
+end
+=#
+
```

Comment:
@kylebd99 I'm ready to merge this, but I'm not sure what to do with this function-based interface. Would you like to keep an interface like this? Should we implement it with the functions in einsum.jl?

Reply:
I think we can ditch this interface for the moment. Going forward, we'll do optimization on the HL language anyways, and it should be easy to identify an einsum-like query in that language if we need to optimize at that level for some reason. From the python interface perspective, any potential einsum frontend will presumably just immediately translate into an  aggregate(mapjoin(...)) query in HL.